torchgeo/losses/ssl.py

```diff
@@ -0,0 +1,41 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""Loss functions self-supervised learning."""
+
+import torch
+import torch.nn.functional as F
+from torch.nn.modules import Module
+
+
+class NTXentLoss(Module):
+    """The Normalized-temperature Cross Entropy Loss (NT-Xent).
+
+    This loss is defined in `'A Simple Framework for Contrastive Learning of
+    Visual Representations' <https://arxiv.org/abs/2002.05709>`_.
+
+    .. versionadded:: 0.5
+    """
+
+    def forward(
+        self, z1: torch.Tensor, z2: torch.Tensor, t: torch.Tensor
+    ) -> torch.Tensor:
+        """Computes the NT-Xent loss.
+
+        Args:
+            z1: batch of vectors of shape (B, D).
+            z2: batch of vectors of shape (B, D).
+            t: temperature value to normalize the similarity vectors
+
+        Returns:
+            ntxent loss
+        """
+        batch_size = z1.shape[0]
+        device = z1.device
+        z1 = F.normalize(z1, dim=-1)
+        z2 = F.normalize(z2, dim=-1)
+        similarity = torch.matmul(z1, z2.T)
+        similarity = similarity * torch.exp(t)
+        targets = torch.arange(batch_size, device=device)
+        loss = F.cross_entropy(similarity, targets)
+        return loss
```

Comment:
I trust that this works, but the MoCo and SimCLR implementations are so different that I want to take some time to go through the math and confirm for myself that all of these are actually equivalent. At one point I searched for other implementations (there are several) and they were all drastically different. Some are more general (support a variable number of positive pairs), faster, and more flexible. For now, as long as it works for the trainers we're planning, it doesn't need to be too fancy.

Reply:
That's a good point. This implementation takes the dot product and then L2 normalized each vector so it's effectively taking the cosine similarity between vectors and then passing that through a cross entropy loss function. The target in this case is the identity matrix which has the assumption that there is only 1 positive pair for each sample in a batch.

Reply:
Tried comparing your implementation vs. the ones from SimCLR and MoCo and miraculously all implementations get different results! Using [this script](https://github.com/microsoft/torchgeo/files/11133316/test.txt), I get:
```
isaac: 1.6350399255752563
simclr: 2.2017433643341064
moco: 0.0
moco v3: 3.2240943908691406
```
Not sure if I copied something wrong, especially for MoCo v1 which has a separate queue that I didn't include and am not sure how to incorporate. We sure figure out why these all have different results.

Reply:
- the MoCo v1 loss in your script assumes the target indicator is at the 0th index not the diagonal which is why you are getting different values than all others (see `labels = torch.zeros(logits.shape[0], dtype=torch.long)` vs `targets = torch.arange(batch_size, device=device)`
- My loss is closest to MoCov3 except they multiply by 2 at the end for some reason (I think we can probably use this loss instead)
- The SimCLR loss implementation does something different with the target labels in this line `pos_mask = self_mask.roll(shifts=cos_sim.shape[0] // 2, dims=0)`. I'm not sure why this happens but it transforms the identity matrix targets to a different type of target matrix which is likely why we end up with different values. If we don't change the target mask and keep it as the identity matrix the loss ends up being `tensor(1.3000e+16)`

Reply:
MoCo is InfoNCE and SimCLR is NT-Xent so that may also explain the difference

Reply:
The MoCo loss you posted normalizes by a temperature so it's basically NT-Xent and has only 1 positive pair per sample in a batch. It's just at the 0th index instead of the diagonal. There's nothing wrong with doing it this way as long as the 0th index in the similarity matrix represents the cosine similarity of the vectors from each augmented pair